services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java

```diff
@@ -430,16 +430,17 @@ private Response processSingleFlowExecutionModel(AuthenticationExecutionModel mo
             if (authUser == null) {
                 throw new AuthenticationFlowException("authenticator: " + factory.getId(), AuthenticationFlowError.UNKNOWN_USER);
             }
-            if (!authenticator.configuredFor(processor.getSession(), processor.getRealm(), authUser)) {
-                if (factory.isUserSetupAllowed() && model.isRequired() && authenticator.areRequiredActionsEnabled(processor.getSession(), processor.getRealm())) {
-                    //This means that having even though the user didn't validate the
-                    logger.debugv("authenticator SETUP_REQUIRED: {0}", factory.getId());
-                    processor.getAuthenticationSession().setExecutionStatus(model.getId(), AuthenticationSessionModel.ExecutionStatus.SETUP_REQUIRED);
-                    authenticator.setRequiredActions(processor.getSession(), processor.getRealm(), processor.getAuthenticationSession().getAuthenticatedUser());
-                    return null;
-                } else {
-                    throw new AuthenticationFlowException("authenticator: " + factory.getId(), AuthenticationFlowError.CREDENTIAL_SETUP_REQUIRED);
-                }
+        }
+        //  Trigger authenticator setup if needed if possible and if allowed
+        if (authUser != null && factory.isUserSetupAllowed() && !authenticator.configuredFor(processor.getSession(), processor.getRealm(), authUser)) {
+            if (model.isRequired() && authenticator.areRequiredActionsEnabled(processor.getSession(), processor.getRealm())) {
+                //This means that having even though the user didn't validate the
+                logger.debugv("authenticator SETUP_REQUIRED: {0}", factory.getId());
+                processor.getAuthenticationSession().setExecutionStatus(model.getId(), AuthenticationSessionModel.ExecutionStatus.SETUP_REQUIRED);
+                authenticator.setRequiredActions(processor.getSession(), processor.getRealm(), processor.getAuthenticationSession().getAuthenticatedUser());
```

Comment:
`processor.getAuthenticationSession().getAuthenticatedUser()` can be replaced with `authUser` .

Reply:
Yep, will do

Reply:
@vanrar68 Thank you. I've checked your update.